shell/shared/grain.js

```diff
@@ -511,6 +511,62 @@ if (Meteor.isClient) {
             currentGrainId + event.data.setPath);
       } else if (event.data.setTitle) {
         Session.set("grainFrameTitle", event.data.setTitle);
+      } else if (event.data.renderTemplate) {
+        // Request creation of a single-use template with a privileged API token.
+        // Why?  Apps should not be able to obtain capabilities-as-keys to
+        // themselves directly, because those can be leaked through an
+        // arbitrary bit stream or covert channel.  However, apps often need a
+        // way to provide instructions to users to copy/paste with some
+        // privileged token contained within.  By providing this templating in
+        // the platform, we can ensure that the token is only visible to the
+        // shell's origin.
+        var call = event.data.renderTemplate;
+        check(call, Object);
+        var rpcId = call.rpcId;
+        check(rpcId, String);
+        var template = call.template;
+        check(template, String);
+        var assignment = undefined;
+        if (call.roleId) {
```

Comment:
I'm not sure I know the meaning of allAccess tokens yet - is that shorthand for "all permissions defined in this app's pkgdef.capnp"?

As for having a particular role rather than allAccess: in the case of Piwik, we'll want to use the token to allow the pages with tracking code to submit pageview events, but probably don't want to grant that token access to view config data, or actual site statistics.  So I think supporting some limits to the permissions granted is necessary.